spring-cloud-consul-discovery/src/main/java/org/springframework/cloud/consul/discovery/TtlScheduler.java

```diff
@@ -77,22 +87,46 @@ public void remove(String instanceId) {
 		this.serviceHeartbeats.remove(instanceId);
 	}
 
-	private class ConsulHeartbeatTask implements Runnable {
+	static class ConsulHeartbeatTask implements Runnable {
 
 		private String checkId;
 
-		ConsulHeartbeatTask(String serviceId) {
+		private TtlScheduler ttlScheduler;
+
+		ConsulHeartbeatTask(String serviceId, TtlScheduler ttlScheduler) {
 			this.checkId = serviceId;
 			if (!this.checkId.startsWith("service:")) {
 				this.checkId = "service:" + this.checkId;
 			}
+			this.ttlScheduler = ttlScheduler;
 		}
 
 		@Override
 		public void run() {
-			TtlScheduler.this.client.agentCheckPass(this.checkId);
-			if (log.isDebugEnabled()) {
-				log.debug("Sending consul heartbeat for: " + this.checkId);
+			try {
+				this.ttlScheduler.client.agentCheckPass(this.checkId);
+				if (log.isDebugEnabled()) {
+					log.debug("Sending consul heartbeat for: " + this.checkId);
+				}
+			}
+			catch (OperationException e) {
+				if (this.ttlScheduler.heartbeatProperties.isReRegisterServiceOnFailure()
+						&& this.ttlScheduler.reRegistrationPredicate.isEligible(e)) {
+					log.warn(e.getMessage());
+					if (this.ttlScheduler.registeredService != null) {
+						if (log.isInfoEnabled()) {
+							log.info("Re-register " + this.ttlScheduler.registeredService);
+						}
+						this.ttlScheduler.client.agentServiceRegister(this.ttlScheduler.registeredService,
```

Comment:
`this.client`?

Reply:
`client` is not visible in `ConsulHeartbeatTask`

Reply:
but isn't that because you changed the class to `static`?

Reply:
yeah I've change it to static because of the testability.
Would you like me to revert to private and use `ReflectionUtils` in `ConsulHeartbeatTaskTests`?

Reply:
No